torch/csrc/autograd/record_function.h

```diff
@@ -28,6 +29,21 @@ enum class TORCH_API RecordScope : uint8_t {
   USER_SCOPE,
   NUM_SCOPES, // must be the last in the list
 };
+
+// Predefined kinds of TLS callbacks
+enum class TORCH_API CallbackKind : uint8_t {
```

Comment:
Why do we now have fixed buckets of callbacks instead of the list? Not saying it's wrong - just curious why this asymmetry. 

If it was a vector - I guess iteration over it would be marginally faster

Reply:
basically originally we had 'pushCallback' 'popCallback'
but with the changes coming when threads freely set and unset thread local observers I think it's too dangerous to have it, it many cases we might be just wrong and pop a wrong callback if somehow the sequence of push/pop happened to be an incorrect bracket sequence (technically possible)

so I'm moving away from push/pop, to push(or add) callback with handle x, and remove callback with handle x

we could technically make 'add callback' function return a unique handle that user then passes into remove, but that would just make internal implementation more complicated (and slower)
and with the approach in the PR we can just use an array of the fixed size and quickly add/remove observers, check if observer is set, iterate over;
this results in stricter interface, but, if we add some callback in the code base we can add a constant in this enum, or we can just use 'DEFAULT' constant for ad hoc callbacks (e.g. in the tests)

Reply:
realized that actually vector is fine - it is thread local, we can take our time during add/remove and we can iterate through it efficiently; we only need to store pairs (callback(s), unique_id) and return this id from push callback